### PR TITLE
JP-3654: NSClean subarray speedup

### LINKED ISSUE #8551: Modify NSClean behaviour for subarray data
_Issue [JP-3654](https://jira.stsci.edu/browse/JP-3654) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

At present, the NSClean step does not work in subarray mode as it tries to handle too much data at the same time and runs out of memory (as opposed to the full array implementation, which goes line by line).  This ticket is to consider a rework of the implementation to allow for future application to subarray data.

See #8547 for some related discussion.

### PR DESCRIPTION
**Pull request by t-brandt**
Resolves [JP-3654](https://jira.stsci.edu/browse/JP-3654)

Closes #8551 

This PR changes the behavior of NSClean in subarray mode.  Previously, the code would fit all unilluminated pixels simultaneously.  This required very large amounts of time and memory for larger subarrays.  This PR changes that behavior to compute the correction in overlapping regions, resulting in large speedups with little change in the results.  The reference pixels are no longer used in the correction, which helps with some ringing artifacts near the detector edges.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] ~updated or added relevant tests~
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/nsclean/lib.py [hunk 1]
```diff
@@ -527,12 +527,18 @@ def clean(self, weight_fit=True):
...
             The cleaned data array.
         """ 
         self.fit(weight_fit=weight_fit)  # Fit the background model
-        self.data -= self.model  # Overwrite data with cleaned data
-        return(self.data)
+        if return_model:
+            return self.model
+        else:
+            self.data -= self.model  # Overwrite data with cleaned data
+            return(self.data)
```
Comment (melanieclarke) [inline jwst/nsclean/lib.py:544]:
```suggestion
            return self.data
```

File: jwst/nsclean/nsclean.py [hunk 2]
```diff
@@ -134,8 +133,11 @@ def create_mask(input_model, mask_spectral_regions, n_sigma):
     # Initialize mask to all True. Subsequent operations will mask
     # out pixels that contain signal.
     # Note: mask will be 3D for BOTS mode data
-    mask = np.full(np.shape(input_model.dq), True)
+    #mask = np.full(np.shape(input_model.dq), True)
```
Comment (melanieclarke) [inline jwst/nsclean/nsclean.py:136]: Remove commented out code, clarify comment

File: jwst/nsclean/nsclean.py [hunk 3]
```diff
@@ -134,8 +133,11 @@ def create_mask(input_model, mask_spectral_regions, n_sigma):
...
 
+    # Similar to above, but don't use reference pixels.
+    mask = (input_model.dq>>31)&1 == 0
```
Comment (melanieclarke) [inline jwst/nsclean/nsclean.py:139]: I'm sure the math works out here, but bit shifting is hard to parse, and this isn't robust to potential changes in DQ flag values.  Can you please revise to use `dqflags.pixel['REFERENCE_PIXEL']`?

File: jwst/nsclean/nsclean.py [hunk 4]
```diff
@@ -257,15 +270,92 @@ def clean_subarray(detector, image, mask):
...
+        
+        # also get four nearest neighbors of flagged pixels
+        mask[1:] = mask[1:]&(~outlier[:-1])
+        mask[:-1] = mask[:-1]&(~outlier[1:])
+        mask[:, 1:] = mask[:, 1:]&(~outlier[:, :-1])
+        mask[:, :-1] = mask[:, :-1]&(~outlier[:, 1:])
+            
+
+    # Used to determine the fitting intervals along the slow scan
+    # direction.  Pre-pend a zero so that we sum_mask[i] is equal
```
Comment (melanieclarke) [inline jwst/nsclean/nsclean.py:294]:
```suggestion
    # direction.  Pre-pend a zero so that sum_mask[i] is equal
```

**Review by melanieclarke — COMMENTED** [review]
Thanks for this!  It looks good to me, just a couple small requests.

I will run regression tests and test this branch locally on some subarray.

It looks like there's a conflict in the change log. Can you please fix it?

**Comment by melanieclarke** [comment]
@t-brandt - I'd like to get this in if we can, since we're closing in on the end of the development period for the current build.  Would you mind if I go ahead and make the last clean up changes?

**Comment by t-brandt** [comment]
Please do go ahead and do the last cleanup.  Thanks!

**Comment by melanieclarke** [comment]
Testing locally, results look similar to previous, and processing time is greatly improved.

It's now possible to process subarray=ALLSLITS data, so I removed the restriction on that data set.

Test data was: 
SUBS200A1: jw02757002001_03104_00001_nrs1_rate.fits
ALLSLITS: jw01128004001_0310g_00003_nrs1_rate.fits

**Comment by melanieclarke** [comment]
Regression tests started here:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1690/

**Comment by melanieclarke** [comment]
@t-brandt - Looking at local regression test results, I see a couple things we should maybe discuss some more.

While the ALLSLITS data runs through nsclean, results are quite poor with mask_spectral_regions=True (the default).  I think this is just the known issue with with overfitting highly masked regions, and the workaround is to run with mask_spectral_regions = False and modify nsigma as needed.  I don't think we need to address that in this PR, but we should consider different default values.

Also, the change to remove forcing the edges to be included in the fit (mask=True) seems to be introducing some edge effects in some cases. For the SUBS200A1 regression test data, it shows up only in the reference columns (they are not NaN in this older input rate data).  For the ALLSLITS regression test data, it shows up in the next couple columns too.

Are the edge issues something we can fix in the subarray correction now?  Or should we revert the change to the reference pixel masking?

Here's what the ALLSLITS regression test data looks like.  Top is nscleaned, bottom is the rate file.

allslits_regression_test

**Comment by melanieclarke** [comment]
For comparison, here's what the same data looks like with better masking (top) and with better masking + the reference edges included in the fit (bottom).

allslits_with_fixes

**Comment by t-brandt** [comment]
Maybe what I/we should do then is to zero out the 1/f correction in the reference pixels?  Right now they are masked, and the correction is extrapolated into those pixels (which is a terrible approximation).  Zeroing out the correction in those pixels is a one or two line change.

**Comment by melanieclarke** [comment]
That sounds worth trying - it's similiar to the way the full frame correction sets the correction to zero if there are no unmasked pixels in the column.  For the ALLSLITS case, the edge effects show up in the first 6 columns, so refpix + 2 real science columns.  Would that address those columns too?

**Comment by t-brandt** [comment]
Ok, I have pushed several small changes to fix the edge effects.  These happened when a large, contiguous part of the area used for the correction was masked for fitting 1/f.  In those cases, the matrices became very poorly conditioned.  The approach now skips rows at the end without any available pixels for fitting 1/f.

**Comment by melanieclarke** [comment]
Local regression tests and spot checks with my other test data sets look good.  Thanks for the fix!

I'll run the full regression test set now and hopefully we can get this merged!

**Comment by t-brandt** [comment]
The overfitting present with the default parameters is because the range of frequencies that NSClean is fitting is too large when a large fraction of the array is missing due to the mask.  This can be fixed by changing these frequencies (see the image below).  If we want to actually implement this is will be a change of a few lines; I can do it tomorrow.  I think the most natural way to implement the change would be to check for allslits mode and mask_spectral_regions set to True, and if both of these are set, change the frequency cutoffs from current the default values.  
comparison

**Comment by melanieclarke** [comment]
@t-brandt - that sounds reasonable to me to include here, since this is the first time ALLSLITS data will be correctable with NSClean, and since it really is unusable with current default parameters.

**Comment by melanieclarke** [comment]
Regression tests restarted here:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1702/

**Comment by melanieclarke** [comment]
@hayescr - tagging you in to this conversation for NIRSpec's information.  

This PR is the fix for subarray performance, which enables ALLSLITS subarrays to be corrected.  The default correction frequencies worked very poorly when spectral regions are masked, so we are modifying the cutoff frequencies for that subarray specifically.  We should discuss at a later date if we want to do something similar for the other modes that frequently get overcorrected, like MOS.

**Review by melanieclarke — APPROVED** [review]
Regression test results look as expected.

There are changes to nsclean for all NIRSpec data in spec2:
- ALLSLITS data is corrected for the first time.
- FS subarray corrections are slightly different due to the different algorithm, and due to removing the forced zeros at the edges of the array.
- Full frame corrections for FS, MOS, and IFU are slightly different due to removing the forced zeros at the edges of the array.

Reviewing results locally, all nsclean corrections are visually similar to previous versions. There are also regression test changes for all products downstream of nsclean, as expected.

All of the remaining failures reported are unrelated.

I think this PR is ready to merge now.  When it is in, I will incorporate the changes into the clean_flicker_noise step (#8669).

**Comment by t-brandt** [comment]
Great, thanks @melanieclarke!  @hayescr and @melanieclarke, I think that the issue with overcorrection in various modes is best solved by exposing a single parameter (provisionally called aggressiveness) to the user.  This would tune the frequencies to be corrected and the coupling of the four outputs.  We could set a default with reasonable behavior, with more aggressive removal appropriate to cases where at least ~half of the array is available as a reference for the correction, and less aggressive removals appropriate to cases where most of the array is unavailable for a correction.  I envision this applying to both subarray and full frame corrections.

**Comment by melanieclarke** [comment]
That sounds good to me @t-brandt - it would be nice to be able to tune that without asking the user to modify frequency cutoffs themselves!

I think we don't yet have a ticket for the aggressiveness parameter - can you please file one, and we'll continue the conversation there?  Time is short on this build, but we should have plenty of time to work on it in the next one.

**Comment by t-brandt** [comment]
Sounds perfect.  The ticket would modify the full frame NSClean behavior and create an aggressiveness parameter.  I will make one.

**Comment by hayescr** [comment]
Thanks @melanieclarke and @t-brandt, agreed that we might want slightly different cutoff frequencies for different cases, perhaps other subarrays when the spectral regions are masked as well.  I had a look at some of Tim's tests of setting this aggressiveness and there were some cases where different cases looked better with different values, so this would definitely be worth investigating.  Feel free to tag me on that once you open it Tim, I'm happy to continue the conversation more there.
